Remove wallet.dat path handling from wallet.cpp, rpcwallet.cpp

ryanofsky opened the pull request:
Get rid of file path handling in wallet application code and move it down to database layer.

There is no change in behavior except for some changed error messages.

Motivation for this change is to make code more understandable, but also to prepare for adding SQLite support in #19077 so SQLite implementation can be contained at the database layer and wallet loading code does not need to become more complicated.

ryanofsky committed:
Remove WalletLocation class

This removes a source of complexity and indirection that makes it harder to
understand path checking code. Path checks will be simplified in upcoming
commits.

There is no change in behavior in this commit other than a slightly more
descriptive error message in `loadwallet` if the default "" wallet can't be
found. (The error message is improved more in upcoming commit "wallet: Remove
path checking code from loadwallet RPC".)

hebasto commented:
Concept ACK.

achow101 commented:
Could the second commit be split up a bit? It seems to be doing a lot and several functions are being renamed and removed.

ryanofsky commented:
> Could the second commit be split up a bit? It seems to be doing a lot and several functions are being renamed and removed.

Sure, I think I can split up like you're suggesting. The main change is introducing the `MakeBerkeleyDatabase` function. Then that function is called various places to get rid of path operations. So I think introducing MakeBerkeleyDatabase could be one commit, and then this can be followed by more commits calling it from RPC code, and then gui code, and then wallet tool code

ryanofsky commented:
Oh, also I didn't want this PR to depend on #19099, but this would be a little smaller with #19099 merged frist, because then there would be no need to change interfaces/node and walletinitinterface code here

luke-jr commented:
I can see why we'd want to move anything about "wallet.dat" or the "wallet is a directory" stuff into the database layer, but IMO it doesn't make sense to hide the path the wallet is located within...

meshcollider commented:
Concept ACK, as a logical separation I think it makes sense to have database-format-specific stuff like this at a lower layer than the logical use of a generic database in the rest of the code.

ryanofsky committed:
wallet: Add MakeDatabase function

New function is not currently called but will be called in upcoming commits. It
moves database path checking, and existence checking, and already-loaded
checking, and verification into a single function so this logic does not need
to be repeated all over higher level wallet code, and so higher level code does
not need to change when SQLite support is added in
https://github.com/bitcoin/bitcoin/pull/19077. This also lets higher level
wallet code make fewer assumptions about the contents of wallet directories.

This commit just adds the new function and does not change behavior in any way.

ryanofsky committed:
refactor: Use DatabaseStatus and DatabaseOptions types

No changes in behavior. Just replaces arguments and return types

ryanofsky committed:
wallet: Remove Verify and IsLoaded methods

Checks are now consolidated in MakeBerkeleyDatabase function instead of
happening in higher level code.

This commit does not change behavior except for error messages which now
include more complete information.

ryanofsky committed:
refactor: Pass wallet database into CWallet::Create

No changes in behavior

ryanofsky committed:
wallet: Remove path checking code from loadwallet RPC

This commit does not change behavior except for error messages which now
include more complete information.

ryanofsky committed:
wallet: Remove path checking code from createwallet RPC

This commit does not change behavior except for error messages which now
include more complete information.

ryanofsky committed:
wallet: Remove path checking code from bitcoin-wallet tool

This commit does not change behavior except for error messages which now
include more complete information.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/19619#issuecomment-665853496

> Could the second commit be split up a bit? It seems to be doing a lot and several functions are being renamed and removed.

Took some work, but I split this up into smaller commits that should make it more obvious nothing is changing.

> I can see why we'd want to move anything about "wallet.dat" or the "wallet is a directory" stuff into the database layer, but IMO it doesn't make sense to hide the path the wallet is located within...

I don't think anything has been hidden here. Database object now has a filename method and error messages are more specific than before including full paths rather than path fragments.

---

Updated 0e4cb50e3c86c3093b18fe2dfb47f563db32afef -> 4c2638281446efc7b23559925866238808a87c71 ([`pr/path.3`](https://github.com/ryanofsky/bitcoin/commits/pr/path.3) -> [`pr/path.4`](https://github.com/ryanofsky/bitcoin/commits/pr/path.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/path.3..pr/path.4)) splitting up later commits and making a few minor cleanups

achow101 commented on src/wallet/bdb.cpp:
```diff
@@ -824,3 +824,50 @@ std::unique_ptr<DatabaseBatch> BerkeleyDatabase::MakeBatch(const char* mode, boo
+
+std::unique_ptr<BerkeleyDatabase> MakeBerkeleyDatabase(const fs::path& path, const DatabaseOptions& options, DatabaseStatus& status, bilingual_str& error)
+{
+    status = DatabaseStatus::SUCCESS;
```
I think the non-BDB specific parts of this should be part of `MakeDatabase`. It would make this easier to handle in #19077

achow101 commented on src/wallet/db.h:
```diff
@@ -142,6 +140,9 @@ class WalletDatabase
     virtual void ReloadDbEnv() = 0;
 
+    /** Return path to main database file for logs and error messages. */
+    virtual std::string Filename() = 0;
```
There is a `m_file_path` variable in `WalletDatabase` that does this. If we add `Filename`, we should remove that.

achow101 commented on src/wallet/db.h:
```diff
@@ -147,9 +147,6 @@ class WalletDatabase
     int64_t nLastWalletUpdate;
 
-    /** Verifies the environment and database file */
-    virtual bool Verify(bilingual_str& error) = 0;
```
It seems like #19077 still needs `Verify`. Even if we don't necessarily use this outside of database specific code paths, I think it's still useful to keep it part of `WalletDatabase`.

achow101 replied:
I think we can require that verify is an online verification function and call it from `MakeDatabase` rather than `MakeBerkeleyDatabase`.

achow101 replied:
The main idea is that the verify is an option for all `WalletDatabase` implementations and always done within `MakeDatabase` so that there isn't an implementation that doesn't do this verification. I think we should be requiring that all `WalletDatabase` implementations have some kind of online verification, and if a database doesn't provide this, perhaps we shouldn't be using it.

ryanofsky reviewed (commented):
Rebased 4c2638281446efc7b23559925866238808a87c71 -> 977bfc5aef22c21ff9612068faf0672bf0f022eb ([`pr/path.4`](https://github.com/ryanofsky/bitcoin/commits/pr/path.4) -> [`pr/path.5`](https://github.com/ryanofsky/bitcoin/commits/pr/path.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/path.4-rebase..pr/path.5)) due to conflict with #19098, and updated with MakeDatabase suggestion

achow101 commented:
ACK 977bfc5aef22c21ff9612068faf0672bf0f022eb

ryanofsky commented:
Rebased 977bfc5aef22c21ff9612068faf0672bf0f022eb -> f190f67ca7b3315c9dc46bf882dbc96af778b607 ([`pr/path.5`](https://github.com/ryanofsky/bitcoin/commits/pr/path.5) -> [`pr/path.6`](https://github.com/ryanofsky/bitcoin/commits/pr/path.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/path.5-rebase..pr/path.6)) due to conflict with #19457

achow101 commented on src/wallet/wallettool.cpp:
```diff
@@ -109,19 +107,14 @@ bool ExecuteWalletToolFunc(const std::string& command, const std::string& name)
             WalletShowInfo(wallet_instance.get());
             wallet_instance->Close();
         }
     } else if (command == "info" || command == "salvage") {
```
Might as well remove this `else if` as it's pointless now.

achow101 commented on src/wallet/salvage.cpp:
```diff
@@ -18,6 +18,13 @@ typedef std::pair<std::vector<unsigned char>, std::vector<unsigned char> > KeyVa
+    options.require_existing = true;
+    options.verify = false;
+    std::unique_ptr<WalletDatabase> database = MakeDatabase(file_path, options, status, error);
+    if (!database) return false;
```
I think we need to close and destroy the `database`. Having it created and potentially open now could interfere with the salvage below. We only want to do this for the path existence check.

achow101 replied:
What I was asking for was to have a `database = nullptr` here so that `database` isn't valid further down.

If we do want to do the salvage through `WalletDatabase` (I'm not sure that we should though), then I would expect that all of this recovery code ends up back in `bdb.cpp/h`.

promag commented on src/wallet/bdb.h:
```diff
@@ -144,6 +144,9 @@ class BerkeleyDatabase : public WalletDatabase
     bool Verify(bilingual_str& error);
 
+    /** Return path to main database filename */
+    std::string Filename() override { return (env->Directory() / strFile).string(); }
```
ee5bb1bb09d9834edb6cd15d233be504f00615db

Could be const?

promag commented on src/wallet/wallet.cpp:
```diff
@@ -3727,23 +3729,23 @@ std::unique_ptr<WalletDatabase> MakeWalletDatabase(const std::string& name, cons
 }
 
-std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain, const std::string& name, bilingual_str& error, std::vector<bilingual_str>& warnings, uint64_t wallet_creation_flags)
+std::shared_ptr<CWallet> CWallet::Create(interfaces::Chain& chain, const std::string& name, std::unique_ptr<WalletDatabase> database, uint64_t wallet_creation_flags, bilingual_str& error, std::vector<bilingual_str>& warnings)
```
ee5bb1bb09d9834edb6cd15d233be504f00615db

Why not take `name` from  `database`? Is it intentional to allow independent names and filenames?

promag reviewed (commented):
Concept ACK.

> Motivation for this change is to make code more understandable, but also to prepare for adding SQLite support in #19077 so SQLite implementation can be contained at the database layer and wallet loading code does not need to become more complicated.

So now we have `MakeWalletDatabase` and then `CWallet::Create(database)` which makes perfect sense to me.

hebasto commented on src/wallet/load.cpp:
```diff
@@ -42,16 +42,16 @@ bool VerifyWallets(interfaces::Chain& chain, const std::vector<std::string>& wal
 
     for (const auto& wallet_file : wallet_files) {
-        WalletLocation location(wallet_file);
+        fs::path path = fs::absolute(wallet_file, GetWalletDir());
```
aac063f28e4ed48180c2ba7a51706af969ca1e55, nit
```suggestion
        const fs::path path = fs::absolute(wallet_file, GetWalletDir());
```

hebasto replied:
aac063f28e4ed48180c2ba7a51706af969ca1e55
Add `#include <fs.h>` ?

hebasto commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2499,21 +2499,22 @@ static UniValue loadwallet(const JSONRPCRequest& request)
 
     WalletContext& context = EnsureWalletContext(request.context);
-    WalletLocation location(request.params[0].get_str());
+    std::string name(request.params[0].get_str());
```
aac063f28e4ed48180c2ba7a51706af969ca1e55, nit
```suggestion
    const std::string name(request.params[0].get_str());
```

hebasto commented:
Commit 134f1807e51a8b64c813bf994c8752d55ff3469f "wallet: Add MakeDatabase function" confuses me:

1. `bool IsBerkeleyBtree()` declaration is added to `wallet/bdb.h` but its definition remains in `wallet/walletutil.cpp`

2. `std::unique_ptr<WalletDatabase> MakeDatabase()` declaration is added to `wallet/db.h` but its definition is added to `wallet/walletdb.cpp`

If this made on purpose, mind explaining such code organization?

hebasto commented on src/wallet/bdb.cpp:
```diff
@@ -824,3 +824,35 @@ std::unique_ptr<DatabaseBatch> BerkeleyDatabase::MakeBatch(const char* mode, boo
+{
+    std::unique_ptr<BerkeleyDatabase> db;
+    {
+        LOCK(cs_db);
```
134f1807e51a8b64c813bf994c8752d55ff3469f
Why locking is required here? Doesn't `GetWalletEnv()` do this job?

ryanofsky reviewed (commented):
Thanks for reviews!

Rebased f190f67ca7b3315c9dc46bf882dbc96af778b607 -> 1233419c88c31f21e1b59b4e3f0cdad34d92413b ([`pr/path.6`](https://github.com/ryanofsky/bitcoin/commits/pr/path.6) -> [`pr/path.7`](https://github.com/ryanofsky/bitcoin/commits/pr/path.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/path.6-rebase..pr/path.7)) due to silent conflict with #15937 and updated with a few code review suggestions.

re: https://github.com/bitcoin/bitcoin/pull/19619#issuecomment-674370405

> Commit [134f180](https://github.com/bitcoin/bitcoin/commit/134f1807e51a8b64c813bf994c8752d55ff3469f) "wallet: Add MakeDatabase function" confuses me:
> 
> [...]
> 
> If this made on purpose, mind explaining such code organization?

This PR is not trying to change code organization, just consolidate path manipulation code in the wallet. But here is the desired organization:

- `db.h`, `db.cpp`: should contain `Database` and `Batch` classes, `DatabaseStatus` and `DatabaseOptions` type definitions, and `MakeDatabase()` function referencing the class and types
- `bdb.h`, `bdb.cpp`: should contain `BerkeleyDatabase` and `BerkeleyBatch` classes and `MakeBerkeleyDatabase()` and `ExistsBerkeleyDatabase()` functions
- `sqlite.h`, `sqlite.cpp`: should contain `SqliteDatabase` and `SqliteBatch` classes and `MakeSqliteDatabase()` and `ExistsSqliteDatabase()` functions
- `walletutil.h`, `walletutil.cpp`: should be removed. `BerkeleyBTree()` should be dropped and inlined or  moved to `bdb.h` / `bdb.cpp`, other functions similarly should be dropped or moved.

I'm ok with changing code organization in this PR if it doesn't increase size of this PR, but otherwise think it should be left to a separate moveonly PR. Otherwise this PR is just trying to change things as little as possible and not make the circular include linter angry, so it is using forward declarations in the interim some places to avoid the need to move code right now.

achow101 commented:
ACK 1233419c88c31f21e1b59b4e3f0cdad34d92413b Since last only rebase and some suggestions. My comments on this aren't really blocking.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2513,24 +2513,25 @@ static UniValue loadwallet(const JSONRPCRequest& request)
-    if (!location.Exists()) {
-        throw JSONRPCError(RPC_WALLET_NOT_FOUND, "Wallet " + location.GetName() + " not found.");
-    } else if (fs::is_directory(location.GetPath())) {
+    if (fs::symlink_status(path).type() == fs::file_not_found) {
```
in commit 83e4ead9917735b2ba92c8a35fe37504206ffe32

please don't change behavior in refactoring commits.

Previously:

```
test_framework.authproxy.JSONRPCException: Wallet  not found. (-18)
```

Now:

```
test_framework.authproxy.JSONRPCException: Directory  does not contain a wallet.dat file. (-18)

MarcoFalke replied:
I sort of liked that the functionality was wrapped in one function. This avoids issues exactly like this

meshcollider commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2514,24 +2514,17 @@ static UniValue loadwallet(const JSONRPCRequest& request)
     std::shared_ptr<CWallet> const wallet = LoadWallet(*context.chain, name, options, status, error, warnings);
-    if (!wallet) throw JSONRPCError(RPC_WALLET_ERROR, error.original);
+    if (!wallet) {
+        RPCErrorCode code = status == DatabaseStatus::FAILED_NOT_FOUND || status == DatabaseStatus::FAILED_BAD_FORMAT ? RPC_WALLET_NOT_FOUND : RPC_WALLET_ERROR;
```
Why treat `FAILED_BAD_FORMAT ` as a not-found error?

meshcollider reviewed (commented):
- [X] 83e4ead9917735b2ba92c8a35fe37504206ffe32 Remove WalletLocation class
- [X] 44f3cf2ba538259ad1f7b02322e5dd96adca1bf1 wallet: Add MakeDatabase function
- [X] 7fdfb3951be14c5ab52f2ddf03a384f13649c00b refactor: Use DatabaseStatus and DatabaseOptions types
- [X] 7c2f3b691727de3d2d488e4cc8aea050c28ba9d1 wallet: Remove Verify and IsLoaded methods
- [X] 66d3c200d4225cd6134ee8ffd385ec6fc7961dd8 refactor: Pass wallet database into CWallet::Create
- [X] dde9986ce70215ba1393194e9ec1955eff3a710b wallet: Remove path checking code from loadwallet RPC
- [X] 668568b453060e13a3cc402507fcbda21d3f6785 wallet: Remove path checking code from createwallet RPC
- [X] 1233419c88c31f21e1b59b4e3f0cdad34d92413b wallet: Remove path checking code from bitcoin-wallet tool

Initial read through looks good to me, modulo MarcoFalke's comment on the first commit. Once rebased I will re-review.

Sjors commented:
Concept ACK. On macOS `wallet_multiwallet.py` fails for me with:

```
2020-08-31T12:20:37.628000Z TestFramework (INFO): Concurrent wallet loading
2020-08-31T12:20:38.768000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/Users/sjors/dev/bitcoin/test/functional/test_framework/test_framework.py", line 118, in main
    self.run_test()
  File "test/functional/wallet_multiwallet.py", line 242, in run_test
    assert_equal(got_loading_error, True)
```

ryanofsky reviewed (commented):
Rebased 1233419c88c31f21e1b59b4e3f0cdad34d92413b -> 83a8f51da5bb86c39f030e027100f0eb5f1f52ad ([`pr/path.7`](https://github.com/ryanofsky/bitcoin/commits/pr/path.7) -> [`pr/path.8`](https://github.com/ryanofsky/bitcoin/commits/pr/path.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/path.7-rebase..pr/path.8)) due to conflicts with #19099 and #19671

For some reason github won't let me reply directly to these comments above:

re: https://github.com/bitcoin/bitcoin/pull/19619#discussion_r478516042
> please don't change behavior in refactoring commits.

Good catch. It wasn't my intention to change behavior in this commit. I think I was paying more attention to the error code than error text. But the new text actually makes a little more sense than the old text, so I just noted the change in the commit message. Both these error messages are replaced anyway in later commit "Remove path checking code from loadwallet RPC".

re: https://github.com/bitcoin/bitcoin/pull/19619#discussion_r478519929

> I sort of liked that the functionality was wrapped in one function. This avoids issues exactly like this

Context for this comment may have been lost so I'm not sure what it is referring to. If it is referring to the removed WalletLocation code in `loadwallet`, the PR is inlining operations that were happening across three functions (`loadwallet`, `WalletLocation::WalletLocation`, `WalletLocation::Exists`) into a straightforward path check done one place..

re: https://github.com/bitcoin/bitcoin/pull/19619#issuecomment-683751130

> Concept ACK. On macOS `wallet_multiwallet.py` fails for me with:

Does the failure happen reliably? This is known to be a flaky check: https://github.com/bitcoin/bitcoin/pull/19300#discussion_r475075615

achow101 commented:
ACK 83a8f51da5bb86c39f030e027100f0eb5f1f52ad

hebasto commented on src/wallet/salvage.cpp:
```diff
@@ -18,6 +18,13 @@ typedef std::pair<std::vector<unsigned char>, std::vector<unsigned char> > KeyVa
+    DatabaseStatus status;
+    options.require_existing = true;
+    options.verify = false;
+    std::unique_ptr<WalletDatabase> database = MakeDatabase(file_path, options, status, error);
```
83a8f51da5bb86c39f030e027100f0eb5f1f52ad, nit:
```suggestion
    auto database = MakeDatabase(file_path, options, status, error);
```

hebasto reviewed (approved):
ACK 83a8f51da5bb86c39f030e027100f0eb5f1f52ad, tested on Linux Mint 20 (x86_64).

ryanofsky reviewed (commented):
Rebased 83a8f51da5bb86c39f030e027100f0eb5f1f52ad -> 7bf6dfbb484adfda3b8df26ee3e2ebda239dd263 ([`pr/path.8`](https://github.com/ryanofsky/bitcoin/commits/pr/path.8) -> [`pr/path.9`](https://github.com/ryanofsky/bitcoin/commits/pr/path.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/path.8-rebase..pr/path.9)) due to conflict with #19754

Thanks for reviews! Reminder to anyone interested that sqlite PR #19077 currently depends on this. And so far achow and hebasto have acked, and meshcollider, promag, and sjors have concept acked this.

achow101 commented:
ACK 7bf6dfbb484adfda3b8df26ee3e2ebda239dd263

hebasto reviewed (approved):
re-ACk 7bf6dfbb484adfda3b8df26ee3e2ebda239dd263

hebasto commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2661,17 +2663,15 @@ static UniValue createwallet(const JSONRPCRequest& request)
-        case WalletCreationStatus::SUCCESS:
-            break;
-        // no default case, so the compiler can warn about missing cases
+    std::shared_ptr<CWallet> wallet = CreateWallet(*context.chain, request.params[0].get_str(), options, status, error, warnings);
```
Ok. Using `auto` is a matter of taste or [philosophy](https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/).

But what is wrong with `const`?

ryanofsky replied:
> Ok. Using `auto` is a matter of taste or [philosophy](https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/).
> 
> But what is wrong with `const`?

I don't have a problem with it, but it isn't my preference to tag every variable that could possibly be tagged const with const. It's clarifying and helpful and some cases, and noise in other cases. In these specific cases, I think `shared_ptr<CWallet>` helps readability, `auto` does not help readability, `const` does not help readability.  If there needs to be more consts, probably better to write a developer guideline and linter than discuss in review comments.

meshcollider reviewed (commented):
Code re-review and functional test run ACK 7bf6dfbb484adfda3b8df26ee3e2ebda239dd263
